c/uca/sifter/unisifeggs.cpp

```diff
@@ -0,0 +1,433 @@
+#include <algorithm>
+#include <array>
+#include <filesystem>
+#include <fstream>
+#include <iostream>
+#include <map>
+#include <ranges>
+#include <string>
+#include <string_view>
+#include <vector>
+
+class CheckFailure {
+ public:
+  ~CheckFailure() {
+    std::terminate();
+  }
+  void operator&(std::ostream&) {}
+};
+
+#define CHECK(condition) \
+  (condition)            \
+      ? (void)0          \
+      : CheckFailure() & std::cerr << ("Check failed: " #condition "\n")
+
+class CodePointRange {
+ public:
+  constexpr CodePointRange(char32_t front, char32_t back)
+      : front_(front), back_(back) {
+    if (empty()) {
+      front_ = 1;
+      back_ = 0;
+    }
+  }
+
+  constexpr bool empty() const {
+    return front_ > back_;
+  }
+
+  constexpr bool contains(char32_t c) const {
+    return c >= front_ && c <= back_;
+  }
+
+  constexpr CodePointRange intersection(CodePointRange const& other) const {
+    return CodePointRange(std::max(front_, other.front_),
+                          std::min(back_, other.back_));
+  }
+
+  constexpr char32_t front() const {
+    return front_;
+  }
+
+  constexpr char32_t back() const {
+    return back_;
+  }
+
+  struct iterator {
+    char32_t operator*() const {
+      return value;
+    }
+
+    void operator++() {
+      ++value;
+    }
+
+    bool operator==(iterator const& other) const = default;
+    bool operator!=(iterator const& other) const = default;
+
+    char32_t value;
+  };
+
+  iterator begin() const {
+    return {front_};
+  }
+
+  iterator end() const {
+    return {back_ + 1};
+  }
+
+ private:
+  char32_t front_;
+  char32_t back_;
+};
+
+constexpr CodePointRange convex_hull(CodePointRange const& r1,
+                                     CodePointRange const& r2) {
+  return CodePointRange(std::min(r1.front(), r2.front()),
+                        std::max(r1.back(), r2.back()));
+}
+
+class CodePointSet {
+ public:
+  // Empty set.
+  CodePointSet() = default;
+
+  bool contains(char32_t c) const {
+    auto const it = std::partition_point(
+        ranges_.begin(), ranges_.end(), [c](CodePointRange const& range) {
+          return range.back() < c;
+        });
+    return it != ranges_.end() && it->contains(c);
+  }
+
+  void Include(CodePointRange range) {
+    auto it = std::partition_point(
+        ranges_.begin(), ranges_.end(), [&range](CodePointRange const& r) {
+          return r.back() < range.front();
+        });
+    if (it == ranges_.end() || it->intersection(range).empty()) {
```

Comment:
optional: The second test could be simpler/faster `range.getEnd() < it->getStart()`, right?

Also, if you were testing `range.getLimit() < it->getStart()` here (limit not end), then the immediately-adjacent case would automatically go to the "else" branch and simplify your "if-then" branch.

Reply:
Looking at it again, this code seems all wrong anyway.

Reply:
Rewrote.